src/execution_plan/optimizations/utilize_indices.c

```diff
@@ -19,136 +23,411 @@ AST_Operator _reverseOp(AST_Operator op) {
-
-		// No filters.
-		if(array_len(filterOps) == 0) continue;
-
-		/* At this point we have all the filter ops (and thus, filter trees) associated
-		 * with the scanned entity. If there are valid indices on any filter and no
-		 * equal or higher precedence OR filters, we can switch to an index scan.
-		 *
-		 * We'll currently use the first matching index, but apply all the filters on
-		 * that property. A later optimization would be to find the index with the
-		 * most filters, or use some heuristic for trying to select the minimal range. */
-
-		int filterOpsCount = array_len(filterOps);
-		for(int i = 0; i < filterOpsCount; i ++) {
-			OpBase *opFilter = filterOps[i];
-			ft = ((Filter *)opFilter)->filterTree;
-			/* We'll only employ indices when we have filters of the form:
-			 * node.property [rel] constant or
-			 * constant [rel] node.property
-			 * If we are not comparing against a constant, then we cannot pre-define useful bounds
-			 * for the index iterator, which diminishes their utility. */
-			lhsType = AR_EXP_GetOperandType(ft->pred.lhs);
-			rhsType = AR_EXP_GetOperandType(ft->pred.rhs);
-			if(lhsType == AR_EXP_VARIADIC && rhsType == AR_EXP_CONSTANT) {
-				filterProp = ft->pred.lhs->operand.variadic.entity_prop;
-				constVal = ft->pred.rhs->operand.constant;
-				op = ft->pred.op;
-			} else if(lhsType == AR_EXP_CONSTANT && rhsType == AR_EXP_VARIADIC) {
-				constVal = ft->pred.lhs->operand.constant;
-				filterProp = ft->pred.rhs->operand.variadic.entity_prop;
-				// When the constant is on the left, reverse the relation in the inequality
-				// to properly set the bounds.
-				op = _reverseOp(ft->pred.op);
-			} else {
-				continue;
-			}
+	// No filters, return.
+	uint filters_count = array_len(filters);
+	if(filters_count == 0) goto cleanup;
 
-			// If we've already selected an index on a different property, continue
-			if(idx && strcmp(idx->attribute, filterProp)) continue;
+	for(uint i = 0; i < filters_count; i++) {
+		Filter *filter = filters[i];
+		FT_FilterNode *filter_tree = filter->filterTree;
 
-			// Try to retrieve an index if one has not been selected yet
-			if(!idx) {
-				idx = GraphContext_GetIndex(gc, label, filterProp);
-				if(!idx) continue;
-				iter = IndexIter_Create(idx, SI_TYPE(constVal));
-			}
+		if(filter_tree->t == FT_N_PRED) {
+			_predicateTreeToRange(filter_tree, string_ranges, numeric_ranges);
+		} else {
+			// OR trees are directly converted into RSQnodes.
+			RSQNode *rsqnode = _filterTreeToQueryNode(filter_tree, rs_idx);
+			rsqnodes = array_append(rsqnodes, rsqnode);
+		}
+	}
 
-			// Tighten the iterator range if possible
-			if(IndexIter_ApplyBound(iter, &constVal, op)) {
-				// Remove filter operations that have been folded into the index scan iterator
-				ExecutionPlan_RemoveOp(plan, opFilter);
-				OpBase_Free(opFilter);
-			}
+	/* Build RediSearch query tree
+	 * Convert each range object to RediSearch query node. */
+	raxIterator it;
+	raxStart(&it, string_ranges);
```

Comment:
Index scans over strings always return empty:
```
$ gq social "MATCH (p:person) WHERE p.name > '' OR p.name < '' RETURN p"
1) 1) "p"
2) (empty list or set)
```
(This comment is attached here randomly; I don't know where the root issue is.)

Reply:
Oh, sorry - this comment was meant to say "index scans with OR filters over strings". Missed some important words!

Reply:
Good catch, might be an issue with RediSearch, checking with the Search team.

Reply:
Was missing a tag node as the parent of either lexrange or token